1/practice/src/main/scala/org/spbsu/mkn/scala/TheGame.scala

```diff
@@ -2,23 +2,123 @@ package org.spbsu.mkn.scala
 
 import scala.io.StdIn.readLine
 import scala.util.Random
+import scala.util.Random
 
 object TheGame {
 
   sealed trait GuessResult
+
   case class Correct(numTries: Int) extends GuessResult
+
   case class Incorrect(bulls: Int, cows: Int) extends GuessResult
 
-  class RepeatingDigitsException extends RuntimeException
-  class WrongNumberLengthException(expected: Int, got: Int) extends RuntimeException
+  case class WrongInput() extends GuessResult
+
+  class RepeatingDigitsException extends RuntimeException {
+    override def getMessage: String = {
+      "Your input contains repeating symbols."
+    }
+  }
+
+  class WrongNumberLengthException(expected: Int, got: Int) extends RuntimeException {
+    override def getMessage: String = {
+      s"Your input length is $got, but expected $expected."
+    }
+  }
+
+  private val alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNPQRSTUVWXYZ0123456789"
+
+  def generateNumberString(length: Int): String = Random.shuffle(alphabet).toString().take(length)
+
+  def validate(secret: String, userInput: String, numTries: Int = 1): GuessResult = {
+    if (secret.length != userInput.length) {
+      throw new WrongNumberLengthException(secret.length, userInput.length)
+    }
+    if (!userInput.forall(ch => userInput.count(_ == ch) == 1)) {
+      throw new RepeatingDigitsException
+    }
+    val secretAsArray = secret.toCharArray
+    if (userInput == secret) {
+      Correct(numTries)
+    }
+    else {
+      Incorrect(
+        (0 until secret.length).count(i => secret(i) == userInput(i)),
+        (0 until secret.length).count(i => secret(i) != userInput(i) && secretAsArray.contains(userInput(i)))
+      )
+    }
+  }
 
-  def generateNumberString(length: Int): String = ???
+  private def readLength(): Int = {
+    var length = 0
+    println("Print preferred length.")
+    try {
+      length = readLine().toInt
+    } catch {
+      case _ =>
+        println("Your input is not a number.")
+        0
+    }
+    length
+  }
 
-  def validate(secret: String, userInput: String, numTries: Int = 1): GuessResult = ???
+  private def getUserInput(secret: String): GuessResult = {
+    val input = readLine()
+    try {
+      validate(secret, input)
+    } catch {
+      case e: RuntimeException =>
+        println(e.getMessage)
+        WrongInput()
+      case _ => ???
```

Comment:
this will swallow all other exceptions and throw an unimplemented exception
you can just omit the last case to let the unhandled exceptions fall through